# Exploring eBay Car Sales Data

## Introduction

In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, [classifieds](https://en.wikipedia.org/wiki/Classified_advertising) section of the German eBay website.

Data dictionary for the dataset:

- **dateCrawled**: When the ad was first crawled. All field-values are taken from this date.
- **name**: Name of the car.
- **seller**: Whether the seller is private or a dealer.
- **offerType**: The type of listing.
- **price**: The listed selling price of the car.
- **abtest**: Whether the listing is included in an A/B test.
- **vehicleType**: The type of vehicle.
- **yearOfRegistration**: The year in which the car was first registered.
- **gearbox**: The type of transmission.
- **powerPS**: The power of the car in PS.
- **model**: The car model name.
- **odometer**: How many kilometers the car has driven.
- **monthOfRegistration**: The month in which the car was first registered.
- **fuelType**: What type of fuel the car uses.
- **brand**: The brand of the car.
- **notRepairedDamage**: If the car has a damage which is not yet repaired.
- **dateCreated**: The date the eBay listing was created.
- **nrOfPictures**: The number of pictures in the ad.
- **postalCode**: The postal code for the location of the vehicle.
- **lastSeenOnline**: When the crawler saw this ad last online.

The aim of this project is to clean the dataset and perform some initial analysis on it. While working on this project, you'll become familiar with some of the unique benefits the JupyterLab environment has to offer for pandas.

In [1]:
import numpy as np
import pandas as pd
import re

In [2]:
autos = pd.read_csv('autos.csv', encoding='latin-1')
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

From the work we did above, we can make the following observations:

- The dataset contains 20 columns, most of which are strings.
- Some columns have null values, but none have more than ~20% null values.
- The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace spaces with underscores.

Let's convert the column names from camelcase to snakecase and reword some of the column names to be more descriptive.

## Cleaning Column Names

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
mapping = {'yearOfRegistration': 'registration_year',
          'monthOfRegistration': 'registration_month',
          'notRepairedDamage': 'unrepaired_damage',
          'dateCreated': 'ad_created'}
autos.rename(mapping, axis=1, inplace=True)
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

*Change the rest of the column names from camelcase to snakecase*

In [6]:
def cleaned_cols(col):
    col = re.sub( '(?<!^)(?=[A-Z])', '_', col ).lower()
    return col
autos.columns = [cleaned_cols(c) for c in autos.columns]
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_p_s', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [7]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

From the descriptive statistics table above, there are a number of things worth noting:

- `price` columns are numeric values stored as text.
- The `seller` and `offer_type` columns contain values where almost all the values are the same, so they can be safely removed from the DataFrame.
- There are some rows of data in the `registration_year` column that can be removed as the table contains a min and max value of 1000 and 9999 which is highly impropable for the vehicle registrations.
- The max in `power_p_s` column is significantly more than the value at 75% which suggest some inaccurate data in this column.
- `odometer` columns are numeric values stored as text.
- The minimum in the `registration_month` column is 0 which is not possible, so this also suggests inaccurate rows of data.
- the `nr_of_pictures` column can be dropped as none of the rows have any pictures.

## Exploring the Odometer and Price Columns

We can also see from previewing the data that the `price` and `odometer` columns are numeric values stored as text so we'll deal with these columns first by:

- Removing any non-numeric characters.
- Convert the column to a numeric dtype.

In [8]:
autos['price'].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

In [9]:
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '').astype(int)
autos['price']

0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int32

In [10]:
autos['odometer']

0        150,000km
1        150,000km
2         70,000km
3         70,000km
4        150,000km
           ...    
49995    100,000km
49996    150,000km
49997      5,000km
49998     40,000km
49999    150,000km
Name: odometer, Length: 50000, dtype: object

In [11]:
autos['odometer'] = autos['odometer'].str.replace(',', '').str.replace('km', '').astype(int)
autos.rename(columns={'odometer': 'odometer_km'}, inplace=True)
autos['odometer_km']

0        150000
1        150000
2         70000
3         70000
4        150000
          ...  
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer_km, Length: 50000, dtype: int32

In [12]:
autos = autos.drop(['nr_of_pictures', 'seller', 'offer_type'], axis=1)

Let's continue exploring the data, specifically looking for data that doesn't look right. We'll start by analyzing the odometer_km and price columns.

In [13]:
print(autos['odometer_km'].unique().shape)
print('----')
print(autos['odometer_km'].describe())
print('----')
print(autos['odometer_km'].value_counts())

(13,)
----
count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
----
odometer_km
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: count, dtype: int64


There are 13 unique rows of data for the odometer_km column. From our obseration we can clearly see that the odometer reading tends to be on the higher side for cars in this dataset. There doesnt seem to be any unrealistically high or low outliers from this column.

In [14]:
print(autos['price'].unique().shape)
print(autos['price'].describe())
print(autos['price'].value_counts().head(15))

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
price
0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
Name: count, dtype: int64


We can see that there are 1421 rows of cars with a price equal 0, we will need to further examine the higher prices of this column

In [15]:
print(autos['price'].value_counts().sort_index(ascending=False).head(10))

price
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: count, dtype: int64


In [16]:
print(autos['price'].value_counts().sort_index(ascending=True).head(15))

price
0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
Name: count, dtype: int64


When looking at the high price range for the column we can see that there is a significant jump up from 350,000 to 999,990 when compared to the steady increase in orices before it. Therefore it is safe to remove any data that is greater than $350,000.

We can also see that there are a number of prices below the 1 prices in our DataFrame, as we know eBay is an auction site and there can be starting bids of $1.

In [17]:
autos = autos[autos['price'].between(1, 350000)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

## Exploring the Date Columns

In [18]:
autos[['date_crawled','ad_created','last_seen', 'registration_year', 'registration_month']][0:10]

date_crawled           ad_created            last_seen  \
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54   
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08   
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37   
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28   
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50   
5  2016-03-21 13:47:45  2016-03-21 00:00:00  2016-04-06 09:45:21   
6  2016-03-20 17:55:21  2016-03-20 00:00:00  2016-03-23 02:48:59   
7  2016-03-16 18:55:19  2016-03-16 00:00:00  2016-04-07 03:17:32   
8  2016-03-22 16:51:34  2016-03-22 00:00:00  2016-03-26 18:18:10   
9  2016-03-16 13:47:02  2016-03-16 00:00:00  2016-04-06 10:46:35   

   registration_year  registration_month  
0               2004                   3  
1               1997                   6  
2               2009                   7  
3               2007                   6  
4               2003                   7  
5               2006                   4  
6               1995                   8  
7               1998                  12  
8               2000                  10  
9               1997                   7

In the DataFrame `date_crawled`, `last_seen` and `ad_created` are columns that are all represented by strings. These will need to be converted to numeric representation so we can analyse it. The `registration_month` and `registration_year` columns are already represented as numeric values.

In [19]:
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

date_crawled
2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: proportion, dtype: float64

We can see that there is uniformity in the distribution of the listings crawled daily.

In [20]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

ad_created
2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: proportion, Length: 76, dtype: float64

We can see the majority of the ads created was within a similar time frame as the crawling dates. The lower frequency of ads created prior to these dates are understandable as once the sales have been completed, the ads can be delisted and won't appear on the website prior to the dates recorded by the crawler

In [21]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()

last_seen
2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: proportion, dtype: float64

There is a spike seen in the distribution for the last three days for the `last_seen` column. This last_seen column tells us of when an ad has been delisted from the site, most likely due to a sale. This high distribution observed during the last 3 days of the crawler recording is most likely not due to a drastic increase in sales as the days prior to these had a roughly similar distribution.

As we previously mentioned, there are rows in the `registration_year` column that are incorrect such as the minimum value of 1000, before cars were invented, and a maximum value of 9999, many years in the future.

We can safely remove any rows with registration years greater than 2016 as this will be definitely inaccurate. To determine the earliest period for the registration of the car is difficult as it could be anywhere during the beginning of the 20th century. To be on the safe side we can see what percentage of vehicles hav registration years outside of 1900 to 2016 and determine if it is a significant number.

In [22]:
autos['registration_year'].between(1900, 2016).sum() / autos['registration_year'].shape[0]

0.961206630289303

We will continue with these two limits and remove the rows outside as the rows outside of them account for less than 1% of the data

In [23]:
autos = autos[autos['registration_year'].between(1900, 2024)]
autos['registration_year'].value_counts(normalize=True, dropna=False).head(10)

registration_year
2000    0.065012
2005    0.060480
1999    0.059677
2004    0.055680
2003    0.055598
2006    0.055001
2001    0.054300
2002    0.051210
1998    0.048676
2007    0.046905
Name: proportion, dtype: float64

We can see that the majority of cars listed in this data were registered in the last 20 years.

Now we'll explore the price of the cars on offer by their brands.

## Exploring Price by Brand

In [24]:
autos['brand'].value_counts(normalize=True)

brand
volkswagen        0.212813
opel              0.108641
bmw               0.108641
mercedes_benz     0.095787
audi              0.085858
ford              0.069667
renault           0.047894
peugeot           0.029457
fiat              0.025996
seat              0.018931
skoda             0.016068
nissan            0.015264
mazda             0.015223
smart             0.014296
citroen           0.014111
toyota            0.012586
hyundai           0.009950
sonstige_autos    0.009641
volvo             0.009043
mini              0.008611
mitsubishi        0.008178
honda             0.007993
kia               0.007107
alfa_romeo        0.006612
porsche           0.005912
suzuki            0.005891
chevrolet         0.005665
chrysler          0.003481
dacia             0.002657
daihatsu          0.002513
jeep              0.002204
subaru            0.002101
land_rover        0.002039
saab              0.001627
daewoo            0.001566
jaguar            0.001524
trabant           0.00

German manufacturers make up the top 5 brands for the cars in the sales data with Volkswagen being the most popular, with approximately double the cars than the next two manufacturers combined.

In [25]:
brand = autos['brand'].value_counts(normalize=True)
most_common_brands = brand[brand>0.04].index
most_common_brands

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford',
       'renault'],
      dtype='object', name='brand')

In [26]:
price_by_brand = {}
for brand in most_common_brands:
    mean_price_by_brand = autos[autos['brand'] == brand]['price'].mean()
    price_by_brand[brand] = int(mean_price_by_brand)

price_by_brand

{'volkswagen': 5333,
 'opel': 2941,
 'bmw': 8261,
 'mercedes_benz': 8526,
 'audi': 9212,
 'ford': 3728,
 'renault': 2431}

We can see that there is a price gap between the top 5 brands in the sales data. We can see that cars manufactured by Audi, BMW and Mercedes Benz tend to be priced higher than the competition. Renault is the least expensive of the top 5 brands while Volkswagen is in between. This could be one of the reasons for the popularity of Volkswagen cars.

We can use a similar method to obtain the average mileage for those cars and then see if there's any link with the mean price. We'll combine both these series objects into a single dataframe, with a shared index, so we can easily compare the two.

In [27]:
brand_mean_mileage = {}
for brand in most_common_brands:
    mean_mileage = autos.loc[autos['brand'] == brand, 'odometer_km'].mean()
    brand_mean_mileage[brand] = int(mean_mileage)
    
brand_mean_mileage

{'volkswagen': 128955,
 'opel': 129452,
 'bmw': 132682,
 'mercedes_benz': 130848,
 'audi': 129492,
 'ford': 124349,
 'renault': 128062}

In [28]:
mean_price = pd.Series(price_by_brand)
mean_mileage = pd.Series(brand_mean_mileage)

In [29]:
top_brand_info = pd.DataFrame(mean_price, columns=['mean_price'])
top_brand_info

mean_price
volkswagen           5333
opel                 2941
bmw                  8261
mercedes_benz        8526
audi                 9212
ford                 3728
renault              2431

In [30]:
top_brand_info['mean_mileage'] = mean_mileage
top_brand_info

mean_price  mean_mileage
volkswagen           5333        128955
opel                 2941        129452
bmw                  8261        132682
mercedes_benz        8526        130848
audi                 9212        129492
ford                 3728        124349
renault              2431        128062

We can see that the car mileage doesn't vary as much as the prices. We can see that the more expensive brands generally tend to have higher mileage with the only outlier being Opel.

Let's continue to clean up the data. To do this we'll indentify the categorical data that uses german words and translate and map the values to thir english counterparts.



In [31]:
autos['gearbox'].value_counts()

gearbox
manuell      36097
automatik    10117
Name: count, dtype: int64

In [32]:
autos['unrepaired_damage'].value_counts()

unrepaired_damage
nein    34771
ja       4689
Name: count, dtype: int64

In [33]:
autos['fuel_type'].value_counts()

fuel_type
benzin     29363
diesel     14352
lpg          668
cng           73
hybrid        37
elektro       19
andere        17
Name: count, dtype: int64

In [34]:
autos['vehicle_type'].value_counts()

vehicle_type
limousine     12598
kleinwagen    10585
kombi          8931
bus            4031
cabrio         3016
coupe          2463
suv            1965
andere          390
Name: count, dtype: int64

In [35]:
autos['model'].value_counts()

model
golf          3898
andere        3438
3er           2686
polo          1688
corsa         1680
              ... 
kappa            2
rangerover       1
i3               1
200              1
b_max            1
Name: count, Length: 245, dtype: int64

In [36]:
gearbox = {'manuell': 'manual', 'automatik': 'automatic'}
unrepaired_damage = {'nein': 'no', 'ja': 'yes'}
fuel_type = {'benzin': 'gasoline', 'elektro': 'electric', 'andere': 'other'}
vehicle_type = {'limousine': 'sedan', 'kleinwagen': 'small car', 'kombi': 'stationwagen', 'cabrio': 'convertible', 'andere': 'other'}
model = {'andere': 'other'}

autos['gearbox'] = autos['gearbox'].map(gearbox)
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(unrepaired_damage)
autos['fuel_type'] = autos['fuel_type'].map(fuel_type)
autos['vehicle_type'] = autos['vehicle_type'].map(vehicle_type)
autos['model'] = autos['model'].map(model)

In [37]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price   abtest  vehicle_type  registration_year    gearbox  power_p_s  \
0   5000  control           NaN               2004     manual        158   
1   8500  control         sedan               1997  automatic        286   
2   8990     test         sedan               2009     manual        102   
3   4350  control     small car               2007  automatic         71   
4   1350     test  stationwagen               2003     manual          0   

   model  odometer_km  registration_month fuel_type       brand  \
0  other       150000                   3       NaN     peugeot   
1    NaN       150000                   6  gasoline         bmw   
2    NaN        70000                   7  gasoline  volkswagen   
3    NaN        70000                   6  gasoline       smart   
4    NaN       150000                   7  gasoline        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0                no  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1                no  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2                no  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3                no  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4                no  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

Now, let's investigate wether or not ther is a price discrepancy if cars that have been previously damaged compared to their non_damaged counterparts.

In [38]:
autos['unrepaired_damage'].value_counts()

unrepaired_damage
no     34771
yes     4689
Name: count, dtype: int64

In [39]:
yes_no = ['yes', 'no']

unrepaired_price = {}
for i in yes_no:
    mean_price = autos[autos['unrepaired_damage'] == i]['price'].mean()
    unrepaired_price[i] = int(mean_price)

unrepaired_price

{'yes': 2221, 'no': 7086}

We can see that cars that have been previously damaged are considerable cheaper than those with no damage prior to their listing.

## Further Analysis

we would use the split the `odometer_km` into groups, and use aggregation to see if average prices follows any patterns based on the mileage.

In [44]:
odometer_groups = ['0-50k (km)', '50k-100k (km)', '100k-150k (km)', '150k+ (km)']
odometer_bins = [0, 50000, 100000, 150000, float('inf')]

autos['odometer_group'] = pd.cut(autos['odometer_km'], bins=odometer_bins, labels=odometer_groups, right=False)
autos['odometer_group']

0            150k+ (km)
1            150k+ (km)
2         50k-100k (km)
3         50k-100k (km)
4            150k+ (km)
              ...      
49995    100k-150k (km)
49996        150k+ (km)
49997        0-50k (km)
49998        0-50k (km)
49999        150k+ (km)
Name: odometer_group, Length: 48545, dtype: category
Categories (4, object): ['0-50k (km)' < '50k-100k (km)' < '100k-150k (km)' < '150k+ (km)']

In [45]:
average_price = autos.groupby('odometer_group')['price'].mean()
average_price

odometer_group
0-50k (km)        14997.381203
50k-100k (km)     10616.856268
100k-150k (km)     6692.709704
150k+ (km)         3728.046603
Name: price, dtype: float64

## Conclusion & Recommendation

In this report, we have been able to clean and analyze the the eBay Car Sales data.

We started by exploring the data to better understand what we are working with and if needed to clean any of the data. It was observed that some fields were not suitable for analysis and as such we had to clean or remove fields based on the information required.

Our initial analysis highlights Volkswagen is notably popular among the featured car advertisements. However, a deeper examination reveals that Citroen, Mercedes-Benz, and Audi secure the top three positions in terms of the highest average prices, while Renault emerges as the least expensive brand.

Further exploration of the data indicates that Citroen boasts both the highest average price and mileage. Conversely, BMW, the fifth-ranked brand, displays higher average mileage compared to its preceding brands, challenging the expected direct correlation between car prices and recorded mileage. Finally, further exploration of the data (amongst other) provided a clear pattern between car mileages and average advertised prices. As mileage increases, there is a corresponding decrease in the average price of cars within the respective mileage groups. Lastly, the data indicates that damaged cars are, on average, cheaper than their undamaged counterparts.